src/Compilers/Core/Portable/SourceGeneration/IncrementalGeneratorRunStep.cs

```diff
@@ -0,0 +1,30 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// A description of a step of an incremental generator that was executed.
+    /// </summary>
+    public sealed class IncrementalGeneratorRunStep
+    {
+        internal IncrementalGeneratorRunStep(string? stepName, ImmutableArray<(IncrementalGeneratorRunStep Source, int OutputIndex)> inputs, ImmutableArray<(object Value, IncrementalStepRunReason OutputState)> outputs, TimeSpan elapsedTime)
+        {
+            Name = stepName;
+            Inputs = inputs;
+            Outputs = outputs;
+            ElapsedTime = elapsedTime;
+        }
+
+        public string? Name { get; }
```

Comment:
>Name [](http://example.com/codeflow?start=23&length=4)

Is this used as a dictionary key? If so, in what scenarios might it be `null`?

Reply:
A step will only have a name when the user provides one with the `WithTrackingName` extension method. Otherwise, the steps produced will not have a name. These nodes will only be accessible through traversing through the graph of steps. Unnamed steps will not be in the NamedSteps dictionary (or the OutputSteps dictionary as those are prenamed).